### PR TITLE
Accept values of the form ``<class name>.<attribute name>`` for the `exclude-protected` list

### LINKED ISSUE #7343: ``exclude-protected`` using the class name and attribute name (astroid's ``qname``) instead of only the attribute name
### Current problem

Right now only the name is taken into account. For example if we want to make ``os._exit`` permissible we'd make all ``x._exit`` permissible. 

### Desired solution

We use the qname instead so we can allow one off class/attr combination.

### Additional context

See https://github.com/PyCQA/pylint/issues/6473

### PR DESCRIPTION
**Pull request by mbyrnepr2**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

Closes #7343

### CODE REVIEW

File: pylint/checkers/classes/class_checker.py [hunk 1]
```diff
@@ -1776,10 +1777,16 @@ def _check_protected_attribute_access(
             Klass.
         """
         attrname = node.attrname
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1779]:
```suggestion
        attrname = node.attrname
        if not is_attr_protected(attrname):
            return
```
Then we can de-indent everything after that.

File: pylint/checkers/classes/class_checker.py [hunk 2]
```diff
@@ -1776,10 +1777,16 @@ def _check_protected_attribute_access(
...
-        if (
-            is_attr_protected(attrname)
-            and attrname not in self.linter.config.exclude_protected
-        ):
+        if is_attr_protected(attrname):
+            if attrname in self.linter.config.exclude_protected:
+                return
+            inferred = safe_infer(node.expr)
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1783]: The inference is costly, might want to do that at the end, after other checks like "it's a typing it's ok".

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks pretty good already, some nit and possible perf improvements.

File: pylint/checkers/classes/class_checker.py [hunk 3]
```diff
@@ -1776,85 +1777,94 @@ def _check_protected_attribute_access(
...
-            # In classes, check we are not getting a parent method
-            # through the class object or through super
-            callee = node.expr.as_string()
+        # If the expression begins with a call to type(self), that's ok.
+        if self._is_type_self_call(node.expr):
+            return
 
-            # Typing annotations in function definitions can include protected members
-            if utils.is_node_in_type_annotation_context(node):
+        inferred = safe_infer(node.expr)
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1815]: Didn't we already infer that line 1784 ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Sorry my refactor suggestion made a mess of the diff, how about we do it in another prior merge  request ?

File: pylint/checkers/classes/class_checker.py (hunk 3 above)
Comment (mbyrnepr2) [reply pylint/checkers/classes/class_checker.py]: Good one - I must have copied the code-added-by-me to this location but not removed it from where it was in the first commit. This commit is fully reverted now until the refactor MR is merged.

File: pylint/checkers/classes/class_checker.py (hunk 2 above)
Comment (mbyrnepr2) [reply pylint/checkers/classes/class_checker.py]: Right good point. However we may need to re-think this method. I realise now that currently it is geared towards class-attributes. If we move this new inference logic below [these lines](https://github.com/PyCQA/pylint/pull/8364/files#diff-28b08eca9b507acd5447a6f7107823b8d5e627963cccc780601da4f88f842813R1805-R1807) then the new logic is never called as that `klass` object tends to be `None`.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Some suggestion about order of execution regarding performance.

File: pylint/checkers/classes/class_checker.py [hunk 4]
```diff
@@ -1777,11 +1778,20 @@ def _check_protected_attribute_access(
         """
         attrname = node.attrname
 
-        if (
+         if (
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1781]:
```suggestion
        if (
```

File: pylint/checkers/classes/class_checker.py [hunk 5]
```diff
@@ -1776,12 +1777,20 @@ def _check_protected_attribute_access(
...
+
+        inferred = safe_infer(node.expr)
+        if (
+            inferred
+            and isinstance(inferred, (nodes.ClassDef, nodes.Module))
+            and f"{inferred.name}.{attrname}" in self.linter.config.exclude_protected
+        ):
+            return
+
         klass = node_frame_class(node)
```
[the remark this answers is not in the document]
Comment (mbyrnepr2) [reply pylint/checkers/classes/class_checker.py:1801]: Sounds good! If you meant to do the refactor in a separate MR let me know. I've made the update in the current MR but that can be undone.

File: pylint/checkers/classes/class_checker.py [hunk 6]
```diff
@@ -1782,21 +1783,31 @@ def _check_protected_attribute_access(
...
+        klass = node_frame_class(node)
         if klass is None:
+            # We are not in a class, no remaining valid case
             self.add_message("protected-access", node=node, args=attrname)
             return
 
+        # In classes, check we are not getting a parent method
+        # through the class object or through super
+        callee = node.expr.as_string()
+
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:1810]: Could we move that to line 1825  ?

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Last perf nit and let's merge.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Comment by mbyrnepr2** [comment]
Thanks @Pierre-Sassoulas. Can we also close https://github.com/PyCQA/pylint/issues/6473 now also?

**Pull request merged**